### PR TITLE
Backend api

### PR DESCRIPTION
**Pull request by mkudlej**
add support for multiple backends
add tests for it
fix some pylint issues

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -133,6 +134,17 @@ def proxy(service, application, api_backend) -> Proxy:
     return proxy
 
 
+@pytest.fixture(scope='module')
+def backendservice(service, backend, application) -> BackendService:
```
Comment (pestanko) [inline tests/integration/conftest.py]: would suggest the name `backend_service`

File: threescale_api/resources.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def __init__(self, *args, entity_name='limit', entity_collection='limits', metri
         self._metric = metric
 
     @property
-    def metric(self) -> 'Metric':
+    def metric(self):
```
Comment (pestanko) [inline threescale_api/resources.py]: why this change?

File: threescale_api/resources.py [hunk 3]
```diff
@@ -637,6 +672,13 @@ def proxy(self) -> 'Proxies':
     def mapping_rules(self) -> 'MappingRules':
         return MappingRules(parent=self, instance_klass=MappingRule)
 
+    @property
+    def oidc(self):
+        return OIDCConfigs(self)
+
+    @property
+    def backendservices(self):
```
Comment (pestanko) [inline threescale_api/resources.py]: I would suggest the name `backend_service`

File: threescale_api/resources.py [hunk 4]
```diff
@@ -637,6 +672,13 @@ def proxy(self) -> 'Proxies':
     def mapping_rules(self) -> 'MappingRules':
         return MappingRules(parent=self, instance_klass=MappingRule)
 
+    @property
+    def oidc(self):
+        return OIDCConfigs(self)
```
Comment (pestanko) [inline threescale_api/resources.py]: here I would provide the type annotation, and also, it is not required to create the instance of the `OIDCConfigs` every time the property is called

File: threescale_api/resources.py (hunk 2 above)
Comment (mganisin) [reply threescale_api/resources.py]: +1

File: threescale_api/resources.py (hunk 4 above)
Comment (mganisin) [reply threescale_api/resources.py]: also comment (at least on github) to explain why this was introduced

**Comment by mkudlej** [comment]
> Also the general comment - I think that some of these tests are too specific and are not testing whether the client is working, more they are testing whether the backend is working, these tests should part of the test suite, not the client implementation.

The tests are similar to tests which are already there for other API functions for system/product and its metrics/methods and so on. I have added just few tests which are special for multi-backend product/service.

File: tests/integration/conftest.py (hunk 1 above)
Comment (mkudlej) [reply tests/integration/conftest.py]: I've renamed it. BackendService(s)->BackendUsage(s) and backendservice(s)->backend_usage(s)

File: threescale_api/resources.py (hunk 2 above)
Comment (mkudlej) [reply threescale_api/resources.py]: There can be Limit linked to Metric linked to Backend. Then it is not Metric, but BackendMetric.

File: threescale_api/resources.py (hunk 3 above)
Comment (mkudlej) [reply threescale_api/resources.py]: Renamed to backend_usage according name of endpoint.

File: threescale_api/resources.py (hunk 4 above)
Comment (mkudlej) [reply threescale_api/resources.py]: I've removed this

File: threescale_api/resources.py [hunk 5]
```diff
@@ -491,6 +498,32 @@ def url(self) -> str:
...
+    def __init__(self, *args, entity_name='metric', entity_collection='metrics', **kwargs):
+        super().__init__(*args, entity_name=entity_name,
+                         entity_collection=entity_collection, **kwargs)
+class BackendMappingRules(MappingRules):
+    def __init__(self, *args, entity_name='mapping_rule', entity_collection='mapping_rules', **kwargs):
+        super().__init__(*args, entity_name=entity_name,
+                         entity_collection=entity_collection, **kwargs)
+
+class BackendServices(Services):
+    def __init__(self, *args, entity_name='backend_usage', entity_collection='backend_usages', **kwargs):
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply threescale_api/resources.py]: Renamed

File: threescale_api/client.py [hunk 6]
```diff
@@ -65,7 +66,7 @@ def admin_api_url(self) -> str:
         """Get admin API url
         Returns(str): URL of the 3scale admin api
         """
-        return self.url + "/admin/api"
+        return f"{self.url}/admin/api"
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply threescale_api/client.py]: reverted

File: threescale_api/resources.py [hunk 7]
```diff
@@ -795,6 +837,37 @@ def app_plan(self) -> ApplicationPlan:
         return self.parent
 
 
+class Backend(DefaultResource):
+    def __init__(self, entity_name='system_name', **kwargs):
+        super().__init__(entity_name=entity_name, **kwargs)
+
+    @property
+    def metrics(self) -> Metrics:
+        return BackendMetrics(parent=self, instance_klass=Metric)
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply threescale_api/resources.py]: Done

File: threescale_api/log_config.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
 import logging
 from logging.config import dictConfig
 
+#LEVEL=logging.DEBUG
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply threescale_api/log_config.py]: reverted

File: threescale_api/resources.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mganisin) [reply threescale_api/resources.py]: Maybe I was misunderstood? look at `instance_klass=Metric` shouldn't it be `instance_klass=BackendMetric` ? Or otherwise should the `BackendMetric` exist at all?

File: threescale_api/resources.py (hunk 2 above)
Comment (jsmadis) [reply threescale_api/resources.py]: @mkudlej ^^

File: threescale_api/resources.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsmadis) [reply threescale_api/resources.py]: @mkudlej ^^
